### PR TITLE
Consolidate RPM identification process

### PR DESCRIPTION
**Pull request by mkosiarc**
Signed-off-by: mkosiarc <mkosiarc@redhat.com>

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- [n/a] JSON/YAML configuration changes are updated in the relevant schema
- [n/a] Changes to metadata also update the documentation for the metadata
- [n/a] Pull request has a link to an osbs-docs PR for user documentation updates
- [n/a] New feature can be disabled from a configuration file

### CODE REVIEW

**Review by ben-alkov — APPROVED** [review]

File: atomic_reactor/plugins/post_rpmqa.py [hunk 1]
```diff
@@ -112,10 +94,9 @@ def gather_output_scratch(self):
             tar_archive.extractall(rpmdb_dir)
 
             rpmdb_path = os.path.join(rpmdb_dir, RPMDB_DIR_NAME)
-            rpmdb_packages = os.path.join(rpmdb_path, RPMDB_PACKAGES_NAME)
 
-            if not os.path.exists(rpmdb_packages):
-                self.log.info('%s does not exist in rpmdb', RPMDB_PACKAGES_NAME)
+            if not os.path.exists(rpmdb_path) or not os.listdir(rpmdb_path):
```
Comment (rcerven) [inline atomic_reactor/plugins/post_rpmqa.py:98]: why are you changing this? this shouldn't be changed

Comment (mkosiarc) [reply atomic_reactor/plugins/post_rpmqa.py:98]: from jira - Fedora: recent fedora have changed names of files, make sure that folder is not empty don't check for exact names.

File: tests/plugins/test_rpmqa.py [hunk 2]
```diff
@@ -163,10 +173,10 @@ def test_rpmqa_plugin(caplog, docker_tasker, base_from_scratch, list_rpms_from_s
         assert workflow.image_components == parse_rpm_output(ignore_autogenerated["package_list"])
 
 
-@pytest.mark.parametrize('packages_exists', [True, False])
```
Comment (rcerven) [inline tests/plugins/test_rpmqa.py:166]: unit test should test file existence for os.listdir

File: atomic_reactor/plugins/post_rpmqa.py (hunk 1 above)
Comment (rcerven) [reply atomic_reactor/plugins/post_rpmqa.py:98]: ack

File: tests/plugins/test_rpmqa.py (hunk 2 above)
Comment (mkosiarc) [reply tests/plugins/test_rpmqa.py:166]: updated the test, I extracted this part to separate testcase to be more straightforward and easier to follow.
You made a very good point as this caught a error when returning None from function gather_output but not from plugin itself.

File: tests/plugins/test_rpmqa.py [hunk 3]
```diff
@@ -163,10 +178,9 @@ def test_rpmqa_plugin(caplog, docker_tasker, base_from_scratch, list_rpms_from_s
...
 @pytest.mark.parametrize('rpm_failed', [True, False])
 @pytest.mark.parametrize('get_archive_raises', [None, APIError, Exception])
-def test_rpmqa_plugin_base_from_scratch(caplog, tmpdir, docker_tasker, packages_exists, rpm_failed,
+def test_rpmqa_plugin_base_from_scratch(caplog, tmpdir, docker_tasker, rpm_failed,
```
Comment (rcerven) [inline tests/plugins/test_rpmqa.py:183]: rename this method, it isn't anymore just for FROM scratch

File: tests/plugins/test_rpmqa.py [hunk 4]
```diff
@@ -221,33 +235,57 @@ def test_rpmqa_plugin_base_from_scratch(caplog, tmpdir, docker_tasker, packages_
...
+            results = runner.run()
+            assert log_msg_getting in caplog.text
+            assert results[PostBuildRPMqaPlugin.key] == PACKAGE_LIST
+            assert workflow.image_components == parse_rpm_output(PACKAGE_LIST)
 
     log_msg = "from scratch, list_rpms_from_scratch is True, trying get rpmdb"
     assert log_msg in caplog.text
 
 
+@pytest.mark.parametrize('rpmdb_dir_exists, rpmdb_dir_empty', [(True, False), (False, False)])
```
Comment (rcerven) [inline tests/plugins/test_rpmqa.py:257]: merge this test with previous test, which was already testing similar thing before

File: tests/plugins/test_rpmqa.py [hunk 5]
```diff
@@ -46,20 +44,25 @@
 pytestmark = pytest.mark.usefixtures('user_params')
 
 
-def generate_archive(tmpdir, empty=False):
+def generate_archive(tmpdir, rpm_dir_exists=True, rpm_dir_empty=False):
     """Generate a fake tar archive
     :param bool empty: whether the archive is empty.
```
Comment (tkdchen) [inline tests/plugins/test_rpmqa.py:49]: docstring is not updated accordingly.

**Review by tkdchen — CHANGES_REQUESTED** [review]

File: tests/plugins/test_rpmqa.py (hunk 5 above)
Comment (mkosiarc) [reply tests/plugins/test_rpmqa.py]: updated

**Comment by rcerven** [comment]
code looks good

tests look fine, but I'd still prefer to have more complex test instead of multiple simple ones, but up to others

**Comment by mkosiarc** [comment]
It really don't think it's a good idea to have complex testcases especially with conditional branching, they are easier to make mistakes in, harder to read/understand and harder to mantain. That's is also one of the reasons we have that story for refactoring atomic-reactor unit-tests (sure there are other issues).
I think unit test should definitely be short, without branching and test only 1 thing if possible, splitting them had additional advantage that I could now name additional testcases and it's clearer what they test, and if they should fail you can debug much quicker where the issue is.

**Comment by lkolacek** [comment]
I can't agree more with Max. Unit test should be small atomic and easy to maintain. The code looks great. Thanks, Max!

**Review by tkdchen — APPROVED** [review]

**Comment by rcerven** [comment]
> It really don't think it's a good idea to have complex testcases especially with conditional branching, they are easier to make mistakes in, harder to read/understand and harder to mantain. That's is also one of the reasons we have that story for refactoring atomic-reactor unit-tests (sure there are other issues).
> I think unit test should definitely be short, without branching and test only 1 thing if possible, splitting them had additional advantage that I could now name additional testcases and it's clearer what they test, and if they should fail we can debug much quicker where the issue is.

yeah problem is that if we will be constantly adding just simple tests, we will end up with tests taking a lot of time

so we should balance between too complex tests and too simple tests

**Comment by mkosiarc** [comment]
> > It really don't think it's a good idea to have complex testcases especially with conditional branching, they are easier to make mistakes in, harder to read/understand and harder to mantain. That's is also one of the reasons we have that story for refactoring atomic-reactor unit-tests (sure there are other issues).
> > I think unit test should definitely be short, without branching and test only 1 thing if possible, splitting them had additional advantage that I could now name additional testcases and it's clearer what they test, and if they should fail we can debug much quicker where the issue is.
> 
> yeah problem is that if we will be constantly adding just simple tests, we will end up with tests taking a lot of time
> 
> so we should balance between too complex tests and too simple tests

Not necessarily, for example when have bigger testcase with many pytest parameters it also has to run for each pytest parameters separately, so when you split it, it will run basically the same code in the same amount of time.

But even if the time necessary for running tests increased slightly, it would still be more beneficial to have more simple testcases as you would save time on understanding/creating/maintaining those tests

**Review by rcerven — APPROVED** [review]

**Comment by MartinBasti** [comment]
Comments addressed here https://github.com/containerbuildsystem/atomic-reactor/pull/1599

**Review by tkdchen — APPROVED** [review]

**Review by MartinBasti — CHANGES_REQUESTED** [review]
Blocking merging - incompatibilities between RPM DB have been identified. This issue must be resolved first.

**Comment by MartinBasti** [comment]
Please add ticket to commit message and provide better description in commit message

**Comment by mkosiarc** [comment]
@MartinBasti I updated the commit message and changed the destination branch, PTAL

**Review by MartinBasti — APPROVED** [review]

**Comment by mkosiarc** [comment]
rn added

**Pull request merged**